conda/common/io.py

```diff
@@ -390,21 +389,48 @@ def close(self):
         self.enabled = False
 
 
-@contextmanager
-def backdown_thread_pool(max_workers=10):
-    """Tries to create an executor with max_workers, but will back down ultimately to a single
-    thread of the OS decides you can't have more than one.
-    """
-    try:
-        yield ThreadPoolExecutor(max_workers)
-    except RuntimeError as e:  # pragma: no cover
-        # RuntimeError is thrown if number of threads are limited by OS
-        log.debug(repr(e))
-        try:
-            yield ThreadPoolExecutor(floor(max_workers / 2))
-        except RuntimeError as e:
-            log.debug(repr(e))
-            yield ThreadPoolExecutor(1)
+class ThreadLimitedThreadPoolExecutor(ThreadPoolExecutor):
+
+    def __init__(self, max_workers=10):
+        super(ThreadLimitedThreadPoolExecutor, self).__init__(max_workers)
+
+    def submit(self, fn, *args, **kwargs):
+        """
+        This is an exact reimplementation of the `submit()` method on the parent class, except
+        with an added `try/except` around `self._adjust_thread_count()`.  So long as there is at
+        least one living thread, this thread pool will not throw an exception if threads cannot
+        be expanded to `max_workers`.
+
+        In the implementation, we use "protected" attributes from concurrent.futures (`_base`
+        and `_WorkItem`). Consider vendoring the whole concurrent.futures library
+        as an alternative to these protected imports.
+
+        https://github.com/agronholm/pythonfutures/blob/3.2.0/concurrent/futures/thread.py#L121-L131  # NOQA
+        https://github.com/python/cpython/blob/v3.6.4/Lib/concurrent/futures/thread.py#L114-L124
+        """
+        with self._shutdown_lock:
+            if self._shutdown:
+                raise RuntimeError('cannot schedule new futures after shutdown')
+
+            f = _base.Future()
+            w = _WorkItem(f, fn, args, kwargs)
+
+            self._work_queue.put(w)
+            try:
+                self._adjust_thread_count()
+            except RuntimeError:
+                # RuntimeError: can't start new thread
+                # See https://github.com/conda/conda/issues/6624
+                if len(self._threads) > 0:
+                    # It's ok to not be able to start new threads if we already have at least
```

Comment:
Ok so this looks like it will prevent a lot of issues, but maybe we can be honest that it will still occur (less likely though). 

Maybe we need to add some python minor/patch version check and add some warning if it is not using the version for which this is tested. Or we could vendor (as@kalefranz  suggest) it if it is small enough. I am just worried we inadvertedly fail due to api changes? 

Maybe we should vendor this as part of this PR already and avoid future issues?

Reply:
> Maybe we should vendor this as part of this PR already and avoid future issues?

Much prefer that to some type of version checking.  I'll vendor it in this PR if @mingwandroid agrees.

Reply:
> Ok so this looks like it will prevent a lot of issues, but maybe we can be honest that it will still occur (less likely though).

Yeah, if *no* threads can be created at all, then we have further work to do.  I don't want to handle that case though until we actually see evidence that it exists.

Reply:
I've looked at code for concurrent.futures in all the various locations.  I'd rather not vendor right now.  We'll of course test conda with python 3.7 before it's released, and other future versions.  Actually, what I *will* do though is add a `run_constrained` requirement for python to conda's recipe.

Reply:
Ok, makes sense